### DISCUSSION TITLE
Backend and device

### DISCUSSION THREAD

**Discussion by wangjiawen2013** [tracel-ai/burn#2417] [Q&A]
Hi,
Burn doesn't explain what "backend" is, assuming the users have already known about it. But for some beginners, they are confused about it. This is adopted from keras documentation:
```text
Keras is a model-level library, providing high-level building blocks for developing deep learning models. It does not handle itself low-level 
operations such as tensor products, convolutions and so on. Instead, it relies on a specialized, well-optimized tensor manipulation library to 
do so, serving as the "backend engine" of Keras. Rather than picking one single tensor library and making the implementation of Keras tied 
to that library, Keras handles the problem in a modular way, and several different backend engines can be plugged seamlessly into Keras.


At this time, Keras has two backend implementations available: the TensorFlow backend and the Theano backend.

[TensorFlow](http://www.tensorflow.org/) is an open-source symbolic tensor manipulation framework developed by Google, Inc.
[Theano](http://deeplearning.net/software/theano/) is an open-source symbolic tensor manipulation framework developed by LISA/MILA Lab at Université de Montréal.
In the future, we are likely to add more backend options. If you are interested in developing a new backend, get in touch!
```
Is the backend definition in burn the same with the above one ? What's the relationship between backend and device ?

**Comment by laggui** [comment]
While it is not explicitly cited, it does state that `the Backend trait [...] enables you to run tensor operations using different implementations without having to modify your code`.

By that definition, it is pretty much in line with the one you gave above.

> What's the relationship between backend and device ?

A backend defines a device type that can be used to perform the operations on (and thus, on which the tensors are allocated). For example, the wgpu backend can use the CPU or GPU (among others), so the device type represents an identifier for each option.
